internal/inventory/azurefetcher/fetcher_storage.go

```diff
@@ -0,0 +1,121 @@
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/cloudbeat/internal/inventory"
+	azurelib "github.com/elastic/cloudbeat/internal/resources/providers/azurelib/inventory"
+)
+
+type storageFetcher struct {
+	logger   *logp.Logger
+	provider storageProvider
+}
+
+type (
+	storageProviderFunc func(context.Context, []azurelib.AzureAsset) ([]azurelib.AzureAsset, error)
+	storageProvider     interface {
+		ListSubscriptions(ctx context.Context) ([]azurelib.AzureAsset, error)
+		ListStorageAccountBlobServices(ctx context.Context, storageAccounts []azurelib.AzureAsset) ([]azurelib.AzureAsset, error)
+		ListStorageAccountQueues(ctx context.Context, storageAccounts []azurelib.AzureAsset) ([]azurelib.AzureAsset, error)
+		ListStorageAccountQueueServices(ctx context.Context, storageAccounts []azurelib.AzureAsset) ([]azurelib.AzureAsset, error)
+		ListStorageAccounts(ctx context.Context, storageAccountsSubscriptionsIds []string) ([]azurelib.AzureAsset, error)
+	}
+)
+
+func newStorageFetcher(logger *logp.Logger, provider storageProvider) inventory.AssetFetcher {
+	return &storageFetcher{
+		logger:   logger,
+		provider: provider,
+	}
+}
+
+func (f *storageFetcher) Fetch(ctx context.Context, assetChan chan<- inventory.AssetEvent) {
+	resourcesToFetch := []struct {
+		name           string
+		function       storageProviderFunc
+		classification inventory.AssetClassification
+	}{
+		{"Storage Blob Services", f.provider.ListStorageAccountBlobServices, inventory.AssetClassificationAzureStorageBlobService},
+		{"Storage Queue Services", f.provider.ListStorageAccountQueueServices, inventory.AssetClassificationAzureStorageQueueService},
+		{"Storage Queues", f.provider.ListStorageAccountQueues, inventory.AssetClassificationAzureStorageQueue},
+	}
+
+	storageAccounts, err := f.listStorageAccounts(ctx)
+	if err != nil {
+		f.logger.Errorf("Could not fetch anything: %v", err)
+		return
+	}
+
+	for _, r := range resourcesToFetch {
+		f.fetch(ctx, storageAccounts, r.name, r.function, r.classification, assetChan)
+	}
+}
+
+func (f *storageFetcher) listStorageAccounts(ctx context.Context) ([]azurelib.AzureAsset, error) {
+	subscriptions, err := f.provider.ListSubscriptions(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error listing subscriptions: %v", err)
+	}
+
+	subscriptionIds := make([]string, len(subscriptions))
+	for _, subscription := range subscriptions {
+		subscriptionIds = append(subscriptionIds, subscription.Name)
+	}
+
+	storageAccounts, err := f.provider.ListStorageAccounts(ctx, subscriptionIds)
+	if err != nil {
+		return nil, fmt.Errorf("error listing storage accounts: %v", err)
+	}
+
+	return storageAccounts, nil
+}
+
+func (f *storageFetcher) fetch(ctx context.Context, storageAccounts []azurelib.AzureAsset, resourceName string, function storageProviderFunc, classification inventory.AssetClassification, assetChan chan<- inventory.AssetEvent) {
+	f.logger.Infof("Fetching %s", resourceName)
```

Comment:
Do we have this for every fetcher? I wonder if we are not over-logging

Reply:
Yes we do. Assuming we'll have ~200 assets, even if we're scanning every minute, that's 400 lines/minute. I'd say that is a very tame amount of logs. If we want to change the approach, then let's do it in a separate PR for all of the fetchers.